### PR TITLE
Add support for retrieving all IdentityFile directives

### PR DESCRIPTION
**Pull request by virtuald**
Additionally, this returns the correct defaults for SSH protocol 2

Unfortunately, since this is the only directive that can have multiple values, there's quite a bit of code duplication. However, it seems to work!

### CODE REVIEW

**Comment by kevinburke** [comment]
I just merged a change to master to fix the Travis branch

**Comment by kevinburke** [comment]
So I didn't anticipate or read the spec closely enough that multiple values could be okay. I don't think that's something that's specific to the IdentityFile directive though, it looks like CertificateFile also allows multiple values, and I stopped looking after that.

I think we need a second method `GetX` and `GetXStrict` that return an array of strings instead of a single string. I tried checking the standard library but the objects that do this generally ask you to "access the map directly" to get the full array (http.Header, url.Values) which won't work in our case.

`GetAll`? `GetMany` ? `List` ?

**Comment by virtuald** [comment]
You're right, I didn't look carefully enough. It seems like the following directives support multiple:

* CertificateFile
* DynamicForward
* RemoteForward
* SendEnv

`GetAll` seems reasonable.

**Comment by kevinburke** [comment]
Please ping when this is ready for re-review!


--
Kevin Burke
925.271.7005 | kev.inburke.com

**Comment by virtuald** [comment]
Updated, tests are good.

**Comment by virtuald** [comment]
One sec, needs a Strict variant. But it'll be mostly the same.

**Comment by virtuald** [comment]
Ok, should be good for review.

**Comment by virtuald** [comment]
Still ready for review. 👍

File: config.go [hunk 1]
```diff
@@ -111,6 +118,18 @@ func Get(alias, key string) string {
 	return DefaultUserSettings.Get(alias, key)
 }
 
+// GetAll retrieves zero or more directives for key for the given alias. GetAll
+// returns nil if no value was found, or if IgnoreErrors is false and we could
+// not parse the configuration file. Use GetStrict to disambiguate the latter
```
Comment (kevinburke) [inline config.go]: s/"Use GetStrict"/"Use GetAllStrict"

File: config.go [hunk 2]
```diff
@@ -273,6 +353,42 @@ func (c *Config) Get(alias, key string) (string, error) {
 	return "", nil
 }
 
+// GetAll returns all values in the configuration that matches the alias and
```
Comment (kevinburke) [inline config.go]: s/that matches/that match

File: config.go [hunk 3]
```diff
@@ -146,6 +186,58 @@ func (u *UserSettings) Get(alias, key string) string {
...
+		case "1":
+			return []string{defaults["identityfile"]}, nil
+		case "2":
+			def := make([]string, len(defaultProtocol2Identities))
+			copy(def, defaultProtocol2Identities)
+			return def, nil
+		}
+	}
+
+	// currently, non-identityfile directives don't have defaults
```
Comment (kevinburke) [inline config.go]: I'm confused about this. If you ask for GetStrict("localhost", "Port") and we return you 22 if there's no match, it seems like we should return you `[]string{"22"}` from `GetAllStrict("localhost", "Port")` if there's no match.

File: config.go [hunk 4]
```diff
@@ -273,6 +353,42 @@ func (c *Config) Get(alias, key string) (string, error) {
 	return "", nil
 }
 
+// GetAll returns all values in the configuration that matches the alias and
+// contains key, or nil if none are present
```
Comment (kevinburke) [inline config.go]: Trailing period here, please.

File: config.go [hunk 5]
```diff
@@ -601,6 +717,25 @@ func (inc *Include) Get(alias, key string) string {
...
+func (inc *Include) GetAll(alias, key string) ([]string, error) {
+	inc.mu.Lock()
+	defer inc.mu.Unlock()
+	// TODO: we search files in any order which is not correct
+	for i := range inc.matches {
+		cfg := inc.files[inc.matches[i]]
+		if cfg == nil {
+			panic("nil cfg")
+		}
+		val, err := cfg.GetAll(alias, key)
```
Comment (kevinburke) [inline config.go]: Is this correct? Should we keep iterating and find all possible values, not just the first match?

Can we change the check below from a nil check to `if len() > 0`? It's possible for a slice to be non-nil and empty.

File: config_test.go [hunk 6]
```diff
@@ -67,6 +67,50 @@ func TestGetWithDefault(t *testing.T) {
...
+	us := &UserSettings{
+		userConfigFinder: testConfigFinder("testdata/identities"),
+	}
+
+	val, err := us.GetAllStrict("hasidentity", "IdentityFile")
+	if err != nil {
+		t.Errorf("expected nil err, got %v", err)
+	}
+	if len(val) != 1 || val[0] != "file1" {
+		t.Errorf("expected [\"file1\"], got %v", val)
```
Comment (kevinburke) [inline config_test.go]: Maybe either

```go
t.Errorf("expected [%q], got %v", "file1", val)
t.Errorf(`expected ["file1"], got %v`, val)
```

File: config_test.go [hunk 7]
```diff
@@ -67,6 +67,50 @@ func TestGetWithDefault(t *testing.T) {
...
+	}
+
+	val, err = us.GetAllStrict("has2identity", "IdentityFile")
+	if err != nil {
+		t.Errorf("expected nil err, got %v", err)
+	}
+	if len(val) != 2 || val[0] != "f1" || val[1] != "f2" {
+		t.Errorf("expected [\"f1\", \"f2\"], got %v", val)
```
Comment (kevinburke) [inline config_test.go]: Same as above.

File: config.go (hunk 5 above)
Comment (virtuald) [reply config.go]: The code was mostly copy/pasted from Get(). Should the nil check be removed there too?

**Comment by virtuald** [comment]
Lost track of this, addressed most concerns.

**Comment by virtuald** [comment]
Ping.

File: validators.go [hunk 8]
```diff
@@ -15,6 +15,26 @@ func Default(keyword string) string {
 	return defaults[strings.ToLower(keyword)]
 }
 
+// DefaultAll returns the default value for the given keyword, but as a slice. If
+// there is no default for the keyword, nil is returned.
+//
+// Some multi-valued settings have different defaults based on other settings, so
+// you must provide the host alias and a function to retrieve a setting
+func DefaultAll(keyword string, alias string, get func(alias, key string) string) []string {
+	if strings.ToLower(keyword) == "identityfile" && get(alias, "Protocol") == "2" {
```
Comment (kevinburke) [inline validators.go]: I'm confused by this function, by this comment: 

```
// Some multi-valued settings have different defaults based on other settings, so
// you must provide the host alias and a function to retrieve a setting
```

and by the fact that this logic appears but not in `Default`. I think maybe if users want this logic they can implement it themselves, alternatively let's pull this out into a different PR.

Comment (virtuald) [reply validators.go]: Originally I had put this logic in GetAll (see ee013b827b53ce170bdc56632a6db6d92eaeefb4). However, I needed to wrap the config, so I need a Default and a DefaultAll. It seems convenient to make it available to people who wanted to do such a thing.

The comment sucks. But if you look at the manpage for ssh_config (OSX 10.12), you can see that the default value for identityfile differs based on which protocol the client has specified. Though, on Linux it seems that this isn't the case. If you want to remove support for legacy stuff, then I can change this to just default to protocol 2.

File: validators.go [hunk 9]
```diff
@@ -160,3 +180,28 @@ var defaults = map[string]string{
...
+var multifileDirectives = map[string]bool{
+	"CertificateFile": true,
+	"IdentityFile":    true,
+	"DynamicForward":  true,
+	"RemoteForward":   true,
+	"SendEnv":         true,
+	"SetEnv":          true,
+}
+
+// IsMultifileDirective indicates a directive that can be found
```
[the remark this answers is not in the document]
Comment (virtuald) [reply validators.go]: From OSX 10.11 for SendEnv:

> Variables are specified by name, which may contain wildcard characters.  Multiple environment variables may be separated by whitespace or spread across multiple SendEnv directives.  The default is not to send any environment variables.

I don't know where SetEnv came from, I can remove it.

[the remark this answers is not in the document]
Comment (virtuald) [reply validators.go]: Actually, [here's where it came from](https://github.com/openssh/openssh-portable/blob/8a85f5458d1c802471ca899c97f89946f6666e61/readconf.c#L283). You can see from the [config parsing source code](https://github.com/openssh/openssh-portable/blob/8a85f5458d1c802471ca899c97f89946f6666e61/readconf.c#L1428) that it will keep adding additional directives as it encounters them.

**Comment by timoreimann** [comment]
I was also trying to retrieve `IdentifyFile`s from my SSH config when I ran into this PR. Is there a chance for the code review to resume progress again?

Also happy to help out myself if that's desirable, though I'd need to ramp up on the project's code base first. @virtuald @kevinburke curious to hear where we stand on this one, thanks!

**Comment by kevinburke** [comment]
Hey, apologies, I didn't realize this was blocked on me. I will add it to my todo list.

**Comment by virtuald** [comment]
Rebased against master again.

**Comment by timoreimann** [comment]
I gave the PR a quick try to retrieve top-level IdentityFiles from my ssh config: it worked like a charm for me. ❤️

Looking forward to seeing this one getting merged.

**Comment by virtuald** [comment]
@kevinburke I've been using this PR for years, works fine. Just merge it, don't let perfect be the enemy of good? :)

**Comment by PaluMacil** [comment]
I'm getting ready to use this library too. This sounds like something I'd appreciate merged. 👍

**Comment by kevinburke** [comment]
I will try to take a look this week

**Comment by virtuald** [comment]
Closing in favor of #28. Also, my repo will be a retargeted fork of this repo with this change so that users don't have to use a replace directive in their module.
